ray_on_golem/server/cluster/sidecars.py

```diff
@@ -0,0 +1,249 @@
+
+    @abstractmethod
+    async def stop(self) -> None:
+        ...
+
+    @abstractmethod
+    async def is_running(self) -> bool:
+        ...
+
+
+class MonitorClusterNodeSidecar(ClusterNodeSidecar, ABC):
+    name = "monitor"
+
+    def __init__(self, node: "ClusterNode", cluster: "Cluster"):
+        super().__init__()
+
+        self._node = node
+        self._cluster = cluster
+
+        self._monitor_task: Optional[asyncio.Task] = None
+
+    @abstractmethod
+    async def _monitor(self) -> None:
+        ...
+
+    async def start(self) -> None:
+        if self.is_running():
+            logger.info(f"Not starting `%s` node {self.name}, as it's already running", self._node)
+            return
+
+        logger.info(f"Starting `%s` node {self.name}...", self._node)
+
+        self._monitor_task = create_task_with_logging(
+            self._monitor(), trace_id=get_trace_id_name(self, self._get_monitor_task_name())
+        )
+
+        logger.info(f"Starting `%s` node {self.name} done", self._node)
+
+    async def stop(self) -> None:
+        if not self.is_running():
+            logger.info(f"Not stopping `%s` node {self.name}, as it's already stopped", self._node)
+            return
+
+        logger.info(f"Stopping `%s` node {self.name}...", self._node)
+
+        await ensure_cancelled(self._monitor_task)
+        self._monitor_task = None
+
+        logger.info(f"Stopping `%s` node {self.name} done", self._node)
+
+    def is_running(self) -> bool:
+        return self._monitor_task and not self._monitor_task.done()
+
+    def _get_monitor_task_name(self) -> str:
+        return "{}-{}".format(self._node, self.name.replace(" ", "-"))
+
+
+class ActivityStateMonitorClusterNodeSidecar(MonitorClusterNodeSidecar):
+    name = "activity monitor"
+
+    def __init__(self, node: "ClusterNode", cluster: "Cluster", check_interval: timedelta) -> None:
+        super().__init__(node, cluster)
+
+        self._check_interval = check_interval
+
+    async def _monitor(self) -> None:
+        from ray_on_golem.server.cluster.nodes import HeadClusterNode
+
+        while True:
+            activity_state = await self._node.activity.get_state()
+
+            if (
+                "Terminated" in activity_state.state
+                or "Unresponsive" in activity_state.state
+                or activity_state.error_message is not None
+            ):
+                logger.warning(
+                    "`%s` node activity is no longer accessible, terminating", self._node
+                )
+
+                provider_desc = await self._node.get_provider_desc(self._node.activity)
+
+                if isinstance(self._node, HeadClusterNode):
+                    self.add_warning_message(
+                        f"Terminating whole cluster as Head({self._node.node_id}) {provider_desc}"
+                        " activity is no longer accessible"
+                    )
+                    create_task_with_logging(self._cluster.stop())
```

Comment:
it seems wrong to me for the node's monitor to signal the cluster to stop... 

this reversal of control seems very hard to maintain in the long run -> if we add such triggers in multiple monitors, it will be cumbersome to change the behavior in case of wider changes... 

rather the other way round, the cluster should detect that its head node is corrupted and should decide to stop. The logic triggering cluster abort should ideally be gathered in one place.

Reply:
That is fair. It should be done at the very beginning.